loopy/statistics.py

```diff
@@ -736,83 +825,143 @@ class MemAccess(ImmutableRecord):
     .. attribute:: kernel_name
 
         A :class:`str` representing the kernel name where the operation occurred.
+
+    .. attribute:: tags
+
+        A :class:`frozenset` of tags to the operation.
     """
 
-    def __init__(self, mtype=None, dtype=None, lid_strides=None, gid_strides=None,
-                 direction=None, variable=None,
-                 *, variable_tags=None,
-                 count_granularity=None, kernel_name=None):
+    address_space: AddressSpace | type[auto] | None = None
+    dtype: LoopyType | None = None
+    lid_strides: Mapping[int, Expression] | None = None
+    gid_strides: Mapping[int, Expression] | None = None
+    read_write: AccessDirection | None = None
+    variable: str | None = None
 
-        if count_granularity not in [*CountGranularity.ALL, None]:
-            raise ValueError("Op.__init__: count_granularity '%s' is "
-                    "not allowed. count_granularity options: %s"
-                    % (count_granularity, [*CountGranularity.ALL, None]))
+    variable_tags: frozenset[Tag] = frozenset()
+    count_granularity: CountGranularity | None = None
+    kernel_name: str | None = None
+    tags: frozenset[Tag] = frozenset()
 
-        if variable_tags is None:
-            variable_tags = frozenset()
+    def __post_init__(self):
+        assert isinstance(self.address_space, (AddressSpace, type(None)))
 
-        if dtype is not None:
+        if self.dtype is not None:
             from loopy.types import to_loopy_type
-            dtype = to_loopy_type(dtype)
+            object.__setattr__(self, "dtype", to_loopy_type(self.dtype))
 
-        ImmutableRecord.__init__(self, mtype=mtype, dtype=dtype,
-                            lid_strides=lid_strides,
-                            gid_strides=gid_strides, direction=direction,
-                            variable=variable, variable_tags=variable_tags,
-                            count_granularity=count_granularity,
-                            kernel_name=kernel_name)
+        if isinstance(self.lid_strides, dict):
+            object.__setattr__(self, "lid_strides", immutabledict(self.lid_strides))
 
-    def __hash__(self):
-        # dicts in gid_strides and lid_strides aren't natively hashable
-        return hash(repr(self))
+        if isinstance(self.gid_strides, dict):
+            object.__setattr__(self, "gid_strides", immutabledict(self.gid_strides))
+
+        if self.variable_tags is None:
+            object.__setattr__(self, "variable_tags", frozenset())
+
+        if not isinstance(self.count_granularity, (CountGranularity, type(None))):
+            raise ValueError(
+                f"unexpected count_granularity: '{self.count_granularity}'")
+
+    @property
+    def mtype(self) -> str:
+        from warnings import warn
+        warn("MemAccess.mtype is deprecated and will stop working in 2024. "
```

Comment:
Bump deadlines to 2026? (here and elsewhere)

Reply:
Done in [0b9d8a4](https://github.com/inducer/loopy/pull/688/commits/0b9d8a468bf9a32fd41fe1009f8060cfab527008)